### PR TITLE
Add support for nonce and message digest for CSP

### PR DESCRIPTION
**Pull request by romainneutron**
Hello,

This PR superseeds the good work done by @kerneldemon in #68 by adding nonce functionality. This PR goes forward and ease the integration by end users by providing Twig integration.

## Message digest

Easily configure message digest algorithm:

```yaml
nelmio_security:
    csp:
        hash:
            algorithm: sha512 # default is sha256, available are sha256, sha384 and sha512
            # Provides CSP Level 1 fallback when using hash or nonce (CSP level 2) 
            # by adding \'unsafe-inline\' source. 
            # See https://www.w3.org/TR/CSP2/#directive-script-src and https://www.w3.org/TR/CSP2/#directive-style-src
            level1_fallback: true # default to true
```

And use it, that's it!
```twig
{% cspscript %}
<script>
    window.api_key = '{{ api_key }}';
</script>
{% endcspscript %}

// ...

{% cspstyle %}
<style>
    body {
        background-color: '{{ bgColor }}';
    }
</style>
{% endcspstyle %}
```

If you're not using Twig:
```php
$listener->addScript("<script>
    window.api_key = '{{ api_key }}';
</script>");


$listener->addStyle("<style>
    body {
        background-color: '{{ bgColor }}';
    }
</style>");
```

## Nonce

Within Twig:
```twig
<script nonce="{{ csp_nonce }}">
    window.api_key = '{{ api_key }}';
</script>

<style nonce="{{ csp_nonce }}">
    body {
        background-color: '{{ bgColor }}';
    }
</style>
```

And without Twig:
```php
// generates a nonce at first time, returns the same nonce once generated
$listener->getNonce();
```

This PR needs a bit more love with:
 - [x] Implement remaining FIX ME
 - [x] Add more tests
 - [x] Update changelog
 - [x] Better compatibility between level 1 and level2

And I'm also wondering about the `cspscript` and `cspstyle`... DO you have better names for these Twig tags?

### CODE REVIEW

File: DependencyInjection/NelmioSecurityExtension.php [hunk 1]
```diff
@@ -136,4 +147,51 @@ public function load(array $configs, ContainerBuilder $container)
             $container->setParameter('nelmio_security.forced_ssl.hosts', $config['forced_ssl']['hosts']);
         }
     }
+
+    private function buildDirectiveSetDefinition($config, $type)
```
Comment (romainneutron) [inline DependencyInjection/NelmioSecurityExtension.php]: For the record @kerneldemon, why did you updated this file, write all this code? Whas it buggy?

Comment (kerneldemon) [inline DependencyInjection/NelmioSecurityExtension.php]: I made this pull request nearly a year ago, so you'll have to excuse me if I don't remember all of the details, but here's the gist of it.

If you look at the older code the `ContentSecurityPolicyListener` has a `fromConfig` factory method that is used to build a directive set for both legacy and up-to-date configuration. 

```php
    public static function fromConfig(array $config)
    {
        if (array_key_exists('report', $config) || array_key_exists('enforce', $config)) {
            $enforce = DirectiveSet::fromConfig($config, 'enforce');
            $report = DirectiveSet::fromConfig($config, 'report');
        } else {
            // legacy config
            $directiveSet = DirectiveSet::fromLegacyConfig($config);

            if ((bool) $config['report_only']) {
                $enforce = new DirectiveSet();
                $report = $directiveSet;
            } else {
                $enforce = $directiveSet;
                $report = new DirectiveSet();
            }
        }

        return new self($report, $enforce, (bool) $config['compat_headers'], $config['hosts'], $config['content_types']);
    }
```

I believe it has no place in this listener.

When adding additional dependencies you are forced to build them in the factory method (the same way DirectiveSet instances are built) or pass them around in multiple places (the constructor and the factory method, including all of its usages). This also complicates injection of optional dependencies that are toggled in the bundle configuration. 

I wanted to inject already built directive sets into the constructor, so I essentially moved the code that does so from the listener. Additionally, I moved the logic from `DirectiveSet::fromLegacyConfig` because the extension is more than capable of handling legacy configuration mapping. I should've removed `getLegacyNamesMap` as well and let the extension do all of the dirty glue work with no pollution in the actual directive class.

I was planning on making further incremental improvements and cleanups after my PR, but got discouraged by the lack of interest for this functionality and ended up using my own fork instead.

TL;DR: most of the code is taken from factory methods.

**Comment by romainneutron** [comment]
Just a reminder for me : there might be a memory leak when using `{% cspscript %}` outside a request context

File: DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -204,6 +218,12 @@ function($v) {
                     $c['report_only'] = false;
                     return $c;
                 })
+            ->end()
+            ->validate()
+                ->ifTrue(function($v) {
+                    return !in_array($v['hash']['algorithm'], array('sha256', 'sha384', 'sha512'), true);
```
Comment (stof) [inline DependencyInjection/Configuration.php]: This is useless as the configuration uses an enumNode which already does exactly this validation

**Comment by Seldaek** [comment]
@romainneutron when you're satisfied with this PR here I'd say go ahead and merge, I won't have time to review in detail but I trust you :)

File: DependencyInjection/Configuration.php (hunk 2 above)
Comment (romainneutron) [inline DependencyInjection/Configuration.php]: oups yes, I forgot I updated with EnumNode, thanks

File: DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -181,6 +181,20 @@ private function addCspNode()
...
+                ->arrayNode('hash')
+                    ->addDefaultsIfNotSet()
+                    ->children()
+                        ->enumNode('algorithm')
+                            ->info('The algorithm to use for hashes')
+                            ->values(array('sha256', 'sha384', 'sha512'))
+                            ->defaultValue('sha256')
+                        ->end()
+                        ->booleanNode('level1_fallback')
+                            ->info('Provide CSP Level 1 fallback when using hash or nonce (CSP level 2) by adding \'unsafe-inline\' source. See https://www.w3.org/TR/CSP2/#directive-script-src and https://www.w3.org/TR/CSP2/#directive-style-src')
```
Comment (romainneutron) [inline DependencyInjection/Configuration.php]: Provide with a "s"

**Comment by romainneutron** [comment]
This PR is now ready for review, this state seems good to me. I give some time to volunteer reviewers before merging this.

**Comment by romainneutron** [comment]
Looks alright to me now, let's merge this, please report any bad behavior folks
